hypothesis-python/tests/nocover/test_precise_shrinking.py

```diff
@@ -0,0 +1,261 @@
+            break
+
+    target_check_value = safe_draw(data, end_marker)
+
+    initial_buffer = bytes(data.buffer)
+
+    replay = ConjectureData.for_buffer(initial_buffer)
+    assert safe_draw(replay, strategy) == initial_value
+    assert safe_draw(replay, end_marker) == target_check_value
+
+    def test_function(data):
+        value = safe_draw(data, strategy)
+        check_value = safe_draw(data, end_marker)
+        if is_interesting(value) and check_value == target_check_value:
+            data.mark_interesting()
+
+    runner = ConjectureRunner(test_function, random=random)
+    try:
+        buf = runner.cached_test_function(initial_buffer)
+        assert buf.status == Status.INTERESTING
+        assert buf.buffer == initial_buffer
+        assert runner.interesting_examples
+        runner.shrink_interesting_examples()
+    except RunIsComplete:
+        assert runner.exit_reason in (ExitReason.finished, ExitReason.max_shrinks)
+    (result,) = runner.interesting_examples.values()
+
+    data = ConjectureData.for_buffer(result.buffer)
+    result_value = safe_draw(data, strategy)
+    data.freeze()
+    return data.as_result(), result_value
+
+
+common_strategies_with_types = [
+    (type(None), st.none()),
+    (bool, st.booleans()),
+    (bytes, st.binary()),
+    (str, st.text()),
+    (int, st.integers()),
+]
+
+common_strategies = [v for _, v in common_strategies_with_types]
+
+
+@lru_cache
+def minimal_buffer_for_strategy(s):
+    return precisely_shrink(s, end_marker=st.none())[0].buffer
+
+
+def test_strategy_list_is_in_sorted_order():
+    assert common_strategies == sorted(
+        common_strategies, key=lambda s: shortlex(minimal_buffer_for_strategy(s))
+    )
+
+
+@pytest.mark.parametrize("typ,strat", common_strategies_with_types)
+@pytest.mark.parametrize("require_truthy", [False, True])
+def test_can_precisely_shrink_values(typ, strat, require_truthy):
+    if typ is type(None) and require_truthy:
+        pytest.skip("None is falsey")
+    if require_truthy:
+        cond = bool
+    else:
+        cond = lambda x: True
+    result, shrunk = precisely_shrink(strat, is_interesting=cond)
+    assert shrunk == find(strat, cond)
+
+
+alternatives = [
+    comb
+    for n in (2, 3, 4)
+    for comb in itertools.combinations(common_strategies_with_types, n)
+]
+
+indexed_alternatives = [
+    (i, j, a) for a in alternatives for i, j in itertools.combinations(range(len(a)), 2)
+]
+
+
+@pytest.mark.parametrize("i,j,a", indexed_alternatives)
+@pytest.mark.parametrize("seed", [0, 4389048901])
+def test_can_precisely_shrink_alternatives(i, j, a, seed):
+    types = [u for u, _ in a]
+    combined_strategy = st.one_of(*[v for _, v in a])
+
+    result, value = precisely_shrink(
+        combined_strategy,
+        initial_condition=lambda x: isinstance(x, types[j]),
+        is_interesting=lambda x: not any(isinstance(x, types[k]) for k in range(i)),
```

Comment:
```suggestion
        is_interesting=lambda x: not any(isinstance(x, types[k]) for k in range(i)),
        seed=seed,
```
or am I confused about python scoping?

Reply:
I think I'm confused about this diff, because I literally can't see any difference between the two lines?

Reply:
The added code is `seed=seed`, I'm not sure how to do a pure insert suggestion in github 😅
(it is possible the `seed` parameter is already getting into the `precisely_shrink` function somehow, I have not like pulled it and added print statements or anything)

Reply:
Oh! Right. Sorry, reading comprehension failure on my part. I think you're right.